airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/thread_based_concurrent_stream.py

```diff
@@ -135,14 +138,39 @@ def _submit_task(self, futures: List[Future[Any]], function: Callable[..., Any],
         futures.append(self._threadpool.submit(function, *args))
 
     def _wait_while_too_many_pending_futures(self, futures: List[Future[Any]]) -> None:
+        self._prune_futures(futures)
+
         # Wait until the number of pending tasks is < self._max_concurrent_tasks
         while True:
-            pending_futures = [f for f in futures if not f.done()]
-            if len(pending_futures) < self._max_concurrent_tasks:
+            if len(futures) < self._max_concurrent_tasks:
                 break
             self._logger.info("Main thread is sleeping because the task queue is full...")
             time.sleep(self._sleep_time)
 
+    def _prune_futures(self, futures: List[Future[Any]]) -> None:
+        """
+        Take a list in input and remove the futures that are completed. If a future has an exception, it'll raise and kill the stream
+        operation.
+
+        Pruning this list thread-safely relies on the assumption that items are only added at the end for the list
+        """
+        if len(futures) < self._max_concurrent_tasks:
+            return
+
+        if self._future_pruning_lock.acquire(blocking=False):
```

Comment:
the futures list is only modified on the main thread so I'm not convinced it needs to be gated by a mutex.